lib/model/unreads.dart

```diff
@@ -0,0 +1,383 @@
+            }
+          case UpdateMessageFlagsRemoveEvent():
+            final newlyUnreadInStreams = <int, Map<String, QueueList<int>>>{};
+            final newlyUnreadInDms = <DmNarrow, QueueList<int>>{};
+            for (final messageId in event.messages) {
+              final detail = event.messageDetails![messageId];
+              if (detail == null) { // TODO(log) if on Zulip 6.0+
+                // Happens as a bug in some cases before fixed in Zulip 6.0:
+                //   https://chat.zulip.org/#narrow/stream/378-api-design/topic/unreads.20in.20unsubscribed.20streams/near/1458467
+                // TODO(server-6) remove Zulip 6.0 comment
+                return;
+              }
+              switch (detail.type) {
+                case MessageType.stream:
+                  ((newlyUnreadInStreams[detail.streamId!] ??= {})[detail.topic!] ??= QueueList())
+                    .add(messageId);
+                case MessageType.private:
+                  final narrow = DmNarrow.ofUpdateMessageFlagsMessageDetail(selfUserId: selfUserId,
+                    detail);
+                  (newlyUnreadInDms[narrow] ??= QueueList())
+                    .add(messageId);
+              }
+            }
+            newlyUnreadInStreams.forEach((incomingStreamId, incomingTopics) {
+              incomingTopics.forEach((incomingTopic, incomingMessageIds) {
+                _addAllInStreamTopic(incomingMessageIds..sort(), incomingStreamId, incomingTopic);
+              });
+            });
+            newlyUnreadInDms.forEach((incomingDmNarrow, incomingMessageIds) {
+              _addAllInDm(incomingMessageIds..sort(), incomingDmNarrow);
+            });
+        }
+    }
+    notifyListeners();
+  }
+
+  // TODO replace with efficient alternative
+  bool _slowIsPresentInStreams(int messageId) {
+    return streams.values.any(
+      (topics) => topics.values.any(
+        (messageIds) => messageIds.contains(messageId),
+      ),
+    );
+  }
+
+  void _addLastInStreamTopic(int messageId, int streamId, String topic) {
+    ((streams[streamId] ??= {})[topic] ??= QueueList()).addLast(messageId);
+  }
+
+  /// [messageIds] must be sorted ascending and without duplicates.
+  void _addAllInStreamTopic(QueueList<int> messageIds, int streamId, String topic) {
+    final topics = streams[streamId] ??= {};
+    QueueList<int> updatedMessageIds = messageIds;
+    if (topics[topic] != null) {
+      // setUnion dedupes existing and incoming unread IDs,
+      // so we tolerate zulip/zulip#22164, fixed in 6.0
+      // TODO(server-6) remove 6.0 comment
+      updatedMessageIds = setUnion(topics[topic]!, messageIds);
+    }
+    topics[topic] = updatedMessageIds;
+  }
+
+  // TODO replace with efficient alternative
+  void _slowRemoveAllInStreams(Iterable<int> messageIds) {
+    for (final messageId in messageIds) {
+      for (final stream in streams.values) {
+        for (final topic in stream.values) {
+          topic.remove(messageId);
```

Comment:
Try this:
```suggestion
  void _slowRemoveAllInStreams(Iterable<int> messageIds) {
    final messageIdSet = Set.of(messageIds);
    for (final stream in streams.values) {
      for (final topic in stream.values) {
        topic.removeWhere((id) => messageIdSet.contains(id));
```

I'm curious how that compares in your timing of mark-as-read events.

That's a more direct transcription of the zulip-mobile version:
```js
function deleteMessages(
  state: UnreadStreamsState,
  ids: $ReadOnlyArray<number>,
): UnreadStreamsState {
  const idSet = new Set(ids);
  const toDelete = id => idSet.has(id);
  const emptyList: Immutable.List<number> = Immutable.List();
  return updateAllAndPrune(state, Immutable.Map(), perStream =>
    updateAllAndPrune(perStream, emptyList, (perTopic: Immutable.List<number>) =>
      perTopic.find(toDelete) ? perTopic.filterNot(toDelete) : perTopic,
    ),
  );
}
```

(The `find ? … :` condition is an optimization to keep the resulting `Immutable.List` identical when it's going to have the same contents; here where we're mutating the same object anyway, that isn't needed.)

If there are N unreads known, and we get an event to remove k of them, then the zulip-mobile version will spend $O(N + k)$ time — $k$ to construct the set, and then $N$ to search through everything and remove matches. But the version in the PR will spend $O(k * N)$ time, which may be substantially more.

Reply:
Could also improve a bit further by having this function take the `Set<int>` as an argument — the caller can reuse it for going through DMs.